### PR TITLE
Reland "PageView scroll physics to match Android #95423"

### LINKED ISSUE #38357: TabView swipe animates too slow compared to Android
## Use case

When you swipe through your tabs, it takes quite some time until the animation is finished. So you really have to wait until you can scroll down. It also slows down even more at the end of the animation. On native android apps this animation is way faster...

## Proposal

Please make the animation similar to android.

Thanks in advance :)

### PR DESCRIPTION
**Pull request by nt4f04uNd**
Fixes #38357
Reland https://github.com/flutter/flutter/pull/95423
Which was reverted to be treated as breaking change here https://github.com/flutter/flutter/pull/97150

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by nt4f04uNd** [comment]
@Piinks is the removal of dependency on `spring` what broke the customer? So a description of a breaking change would be something like "PageViewPhysics now only use ScrollPhysics.spring when going out of range (for example on iOS), while previously it could be used to change the whole animation from user swipes".

**Comment by Piinks** [comment]
Thanks for re-opening! Maybe that was it, what I saw in the customer tests were golden file images. They were capturing pictures of the screen after swiping through pages, but instead of being on the last page as expected, they were on the first page.
The swiping in the test was engaged with a fling gesture. I saw this exact same thing happen across two different customers. WDYT? I can try to get more context if you need it.

**Comment by nt4f04uNd** [comment]
Thanks for the details, it might be the case that this is because of the change of `minFlingDistance` and/or `minFlingVelocity` to beigger values. More context would certainly be helpful.

**Comment by Piinks** [comment]
Customer A

Test code to swipe to next page:

```dart
await fling(customerApp, Offset(-200, 0), 200);
```

I do not know if it makes a difference, but the screen ratio of this test is like a phone in landscape mode, longer across than from top to bottom.

---

Customer B

Same issue, swipes did not end up on the expected page index. Screen size is similar to portrait oriented phone. Swipe gesture from test:

```dart
await tester.fling(
        customerWidget,
        Offset(-200, 0),
        200,
      );
```

---

I think the way to proceed here is to add a temporary opt-in flag. I cannot change the customer tests ahead of time, because the behavior would not be right without this change.

What we usually do here is
1. we add a temporary flag,
2. land the change,
3. migrate the affected customers to the new behavior using the opt in flag, 
4. change the default of the opt in flag to make it an opt out flag
5. update customer code to stop using the flag
6. remove flag
7. release a migration guide documenting the breaking change

I think the ScrollBehavior might be the easiest way to do it, that way I can just opt the whole customer's app in pretty easily, and update the tests, but I'd like to know what you think about all this. Let me know and we can find a way for this to move forward. :)

**Comment by nt4f04uNd** [comment]
(I do not forget about this and my other PRs, it's just that I don't have much time for open source now, anyway I'll try my best to move this forward)

**Comment by Piinks** [comment]
Hey @nt4f04uNd, no worries at all. I really appreciate all of your contributions, they are often non-trivial solutions and very valuable to everyone - so thanks!
If you like, I can pick this one up and work with the affected customers to re-land it. Let me know.

Either way we'll still be here ready to work together when you come back. 😉

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 1]
```diff
@@ -545,13 +560,24 @@ class PageScrollPhysics extends ScrollPhysics {
...
+    return page.roundToDouble();
+  }
+
+  double _getPageDelta(ScrollMetrics position, Tolerance tolerance, double velocity) {
+    final double page = _getPage(position);
+    final double targetPage = _getTargetPage(page, tolerance, velocity);
+    return targetPage - page;
+  }
+
+  double _getTargetPixels(ScrollMetrics position, Tolerance tolerance, double velocity) {
```
Comment (nt4f04uNd) [inline packages/flutter/lib/src/widgets/page_view.dart:577]: @Piinks I just thought whether would it make sense to make some of these methods public, so people can restore the old implementation if they want (or just change it to whatever else)?

**Comment by nt4f04uNd** [comment]
> If you like, I can pick this one up and work with the affected customers to re-land it. Let me know

Thanks, I'd appreciate if you taken this part.

File: packages/flutter/lib/src/widgets/page_view.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/page_view.dart:577]: I don't know if that would be super beneficial, ultimately the changes you are introducing are the way it should be, there are just a lot of tests out there that expect the old behavior.

**Comment by Piinks** [comment]
> > If you like, I can pick this one up and work with the affected customers to re-land it. Let me know
> 
> Thanks, I'd appreciate if you taken this part.

Just confirming, I am going to close this PR and clone it to re-open another one that I can manage landing with affected customers?

**Comment by nt4f04uNd** [comment]
Yeah, sure 🙂

File: packages/flutter/test/widgets/scroll_physics_test.dart [hunk 2]
```diff
@@ -104,13 +104,25 @@ void main() {
...
     const ClampingScrollPhysics clamp = ClampingScrollPhysics();
-    const PageScrollPhysics page = PageScrollPhysics();
 
+    // Verify creating these simulations doesn't alter the velocity for time 0.
+    //
     // Calls to createBallisticSimulation may happen on every frame (i.e. when the maxScrollExtent changes)
-    // Changing velocity for time 0 may cause a sudden, unwanted damping/speedup effect
-    expect(bounce.createBallisticSimulation(position, 1000)!.dx(0), moreOrLessEquals(1000));
-    expect(clamp.createBallisticSimulation(position, 1000)!.dx(0), moreOrLessEquals(1000));
-    expect(page.createBallisticSimulation(position, 1000)!.dx(0), moreOrLessEquals(1000));
```
Comment (Piinks) [inline packages/flutter/test/widgets/scroll_physics_test.dart:113]: This being removed and the comments below it seem like the other issue should really be resolved first.

**Review by Piinks — COMMENTED** [review]
Hey @nt4f04uNd you recently reached out and asked if this was still on my radar. It is, although I have not had the cycles to work on it actively. There are some related issues that need resolution before we can re-land this, as noted in the comments of the test you wrote below.

I also think this should be only applied when the TargetPlatform is Android. The changes here are specific to Android and as you previously established (https://github.com/flutter/flutter/issues/97064), platforms like iOS have their own unique behavior. Especially since this broke customer tests, I think it would be prudent to only apply the change to the platform it is intended to reflect. Applying it to every platform would break more customers - who may be  broken again later when we adjust the behavior for other platforms, creating a lot of churn for users.

**Comment by Piinks** [comment]
Cross-referencing #11599 to maintain link for next planning session.

**Pull request closed without merging**